torch/autograd/grad_mode.py

```diff
@@ -31,13 +32,27 @@ def _wrap_generator(self, func):
         @functools.wraps(func)
         def generator_context(*args, **kwargs):
             gen = func(*args, **kwargs)
-            while True:
-                try:
-                    with self.__class__():
-                        x = next(gen)
-                    yield x
-                except StopIteration:
-                    break
+            grad_mode_context = type(self)()
+            try:
+                with grad_mode_context:
+                    response = gen.send(None)
+                while True:
+                    try:
+                        request = yield response
+                    except GeneratorExit:
+                        with grad_mode_context:
+                            gen.close()
+                        raise
+                    except BaseException:
+                        typ, val, tb = sys.exc_info()
+                        with grad_mode_context:
+                            response = gen.throw(typ, val, tb.tb_next)
+                    else:
+                        with grad_mode_context:
+                            response = gen.send(request)
+            except StopIteration as e:
+                return e.value
```

Comment:
Can the stop iteration have a value here? The previous code was always returning None here.

Reply:
StopIteration is a special exception that is guaranteed to have a [value](https://docs.python.org/3/library/exceptions.html#StopIteration). Quoting from the docs
> When a generator or coroutine function returns, a new StopIteration instance is raised, and the value returned by the function is used as the value parameter to the constructor of the exception.

Therefore, I think, for completeness we should propagate the `.value` as well.

Reply:
Interesting. So the previous code was wrong to ignore that value? Do you have a python code sample that would show how this value is propagated?

I am trying to think if this PR is going to be BC-breaking for any important usage of the current code? Because functions wrapped in this decorator where already returning the right value on exit right?

Reply:
I would not call it "wrong", since haven come across any implementation in which the generator explicitly communicates with the caller through a StopIteration value, i.e. by `return something`. Nevertheless, python specifications state that such behaviour is allowed and documented.

I completely agree, that any modification of the generator behaviour MUST be 100% backwards compatible, so extra care must be taken. However, after studying the docs and the relevant PEPs I cam to the conclusion, that this enhanced generator protocol is compatible with the common `for` [expressions](https://docs.python.org/3/reference/compound_stmts.html#the-for-statement) and the `next` [function](https://docs.python.org/3/library/functions.html#next).